src/main/java/org/opensearch/geospatial/index/mapper/xypoint/XYPointParser.java

```diff
@@ -71,101 +78,158 @@ public static XYPoint parseXYPoint(Object value, final boolean ignoreZValue) thr
+            if (i != 0) {
+                parser.nextToken();
             }
-            element++;
-            if (element == 1) {
-                x = subParser.doubleValue();
-            } else if (element == 2) {
-                y = subParser.doubleValue();
-            } else if (element == 3) {
-                XYPoint.assertZValue(ignoreZValue, subParser.doubleValue());
+
+            if (parser.currentToken() != XContentParser.Token.FIELD_NAME) {
+                if (hasTypePoint == false) {
+                    throw new OpenSearchParseException("field [{}] missing", GEOJSON_TYPE);
+                }
+                if (hasCoordinates == false) {
+                    throw new OpenSearchParseException("field [{}] missing", GEOJSON_COORDS);
+                }
+            }
+
+            if (GEOJSON_TYPE.equals(parser.currentName())) {
+                if (parser.nextToken() != XContentParser.Token.VALUE_STRING) {
+                    throw new OpenSearchParseException("{} must be a string", GEOJSON_TYPE);
+                }
+
+                // To be consistent with geo_shape parsing, ignore case here as well.
+                if (ShapeType.POINT.name().equalsIgnoreCase(parser.text()) == false) {
+                    throw new OpenSearchParseException("{} must be Point", GEOJSON_TYPE);
+                }
+                hasTypePoint = true;
+            } else if (GEOJSON_COORDS.equals(parser.currentName())) {
+                if (parser.nextToken() != XContentParser.Token.START_ARRAY) {
+                    throw new OpenSearchParseException("{} must be an array", GEOJSON_COORDS);
+                }
+                parseXYPointArray(parser, point, ignoreZValue);
+                hasCoordinates = true;
             } else {
-                throw new OpenSearchParseException("[xy_point] field type does not accept more than 3 dimensions");
+                throw new OpenSearchParseException(ERR_MSG_INVALID_FIELDS);
             }
         }
-        return point.reset(x, y);
+
+        return point;
+    }
+
+    private static XYPoint parseXYPointArray(final XContentParser parser, final XYPoint point, final boolean ignoreZValue)
+        throws IOException {
+        try (XContentSubParser subParser = new XContentSubParser(parser)) {
+            double x = Double.NaN;
+            double y = Double.NaN;
+
+            int element = 0;
+            while (subParser.nextToken() != XContentParser.Token.END_ARRAY) {
+                if (parser.currentToken() != XContentParser.Token.VALUE_NUMBER) {
+                    throw new OpenSearchParseException("numeric value expected");
+                }
+                element++;
+                if (element == 1) {
+                    x = parser.doubleValue();
+                } else if (element == 2) {
+                    y = parser.doubleValue();
+                } else if (element == 3) {
+                    GeoPoint.assertZValue(ignoreZValue, parser.doubleValue());
```

Comment:
is user ask not to ignore Z value what we are doing in that case?

Reply:
It will throw exception. Let me avoid to use GeoPoint method here.